# Proyecto: Análisis de datos con Python. Equipo 11
#### Integrantes: 
- ##### Wendy Areli Ramírez Burgos
- ##### Gustavo Corona Ojeda
- ##### Carlos Fernando Vásquez Guerra

**Planteamiento del problema.** 

La ciencia de datos se presenta como una herramienta valuable para el desarrollo de las empresas debido a la tendencia de toma de decisiones guiada por datos, por lo que ésta profesión se perfila como una de gran importancia para el crecimiento de las compañías y de la ciencia en general. 

Actualmente se ha presentado un auge  para el aprendizaje y ejercicio de las herramientas provenientes de la programación y específicamente de la ciencia de datos, pues existe la creencia de que podría representar la oportunidad de una mejora en las expectativas de sueldos y por ende, calidad de vida. 

Al indagar  someramente , ésta aseveración parece ser cierta, sin embargo ¿Qué tan factible es desarrollarse en una carrera en ciencia de datos y cuales son los requerimientos específicos para tener mejores oportunidades? 
Por otro lado, al  ser un campo interdisciplinario que requiere de múltiples habilidades y conocimientos, no existe un "camino" claro que permita comprender las características específicas que debe tener un científico de datos para tener éxito dentro de ésta área laboral. 

**Estado del arte.** 

Llamado por la publicación del Harvard Business Review del  2010 como "El empleo más sexy del siglo XXI" el de  Ciencia de datos se ha posicionado como uno de los más buscados (por empleadores y empleados)  en la última década.  Ya glassdor en el 2016 lo posicionaba como el mejor empleo en los Estados Unidos en 2016 y 2017 y se vio un incremento de un 15% de empleos en ésta área en 2020. 

Según Ramírez (2021) el análisis de datos es una de las profesiones emergentes que será de las más importantes en los próximos 4 años, debido a la necesidad de aprovechar la información que generan los más de 40, 000 millones de dispositivos conectados, lo que ha llevado a las empresas a digitalizar sus procesos para seguir siendo competitivas y, por ende, a buscar perfiles capaces de traducir ésos datos en información de valor para su toma de decisiones, por otro lado, se prevee además que la demanda de profesionales relacionados a las áreas STEM (Science - Technology - Engineering - Mathematics), se incremente en un 12.5% para 2024, según las cifras de la oficina de Estadísticas Laborales de Estados Unidos. 

A pesar de ésta tendencia, se ha encontrado que alrededor del 9% de vacantes disponibles quedan sin cubrir por falta de perfiles especializados, lo que contribuye a que sea una de las profesiones mejor pagadas (Ramírez, 2021). A éste respecto, un estudio realizado por Gouviera (2021) cuyo objetivo era revalar cuáles eran los salarios de un científico de datos en países de habla hispana, encontró que el sueldo promedio de un científico de datos en México es de 29 mil pesos mexicanos al mes e iniciando un promedio de 20 mil al mes, en Monterrey de 16 a 34 mil pesos mexicanos. 

De acuerdo al popular sitio especializado en la intustria STEM, Glassdoor, el puesto como cientítico de datos ocupa el tercer lugar  en América con un sueldo promedio de 120,000 dólares anuales y 1071 ofertas laborales para éste año, 2022; por otro lado, de acuerdo a LinkedIn , el de análisis de datos, es una de las 15 profesiones emergentes más importantes del futuro y se ha convertido en una de las profesiones mejor pagadas debido a la falta de personas que puedan cubrir los requerimientos de éstos puestos (Ramírez, 2021). 

En un estudio realizado por Solís (2021) se concluye que el país con más alta demanda de Científicos de datos, es Estados unidos, con 86,839 vacantes de empleo, seguido de China que cuenta con 42,216, India con 19,578 y Alemania, con 8,581; con respecto a Latinoamérica, se encontró que los países con  con más demanda de Científicos de datos  son Brasil con 3,558 posiciones, apenas un 4% respecto a la demanda de Estados Unidos, seguido de México con 1,465, Argentina con 751, Colombia con 531, Chile con 329 y Perú, con 117. Podemos concluir, por tanto, que ésta tendencia respecto a la demanda y salario atractivo de científicos de datos se presenta también en Latinoamérica, pues los salarios de científico de datos son elevados respecto al promedio (Ramírez, 2018). 

Respecto al  perfil específico de un científico de datos, se ha encontrado que cada compañía tiene su propia definición de "Científico de datos" siendo diversos los requerimientos respecto al manejo de herramientas, habilidades y antecedentes académicos requeridos para desarrollarse en ésta área profesional, pues el rol del científico de datos aún se encuentra en evolución y no se adhiere a un solo conjunto de habilidades (Shaik, 2020). A éste respecto, el estudio de Ramírez (2018), señala que, para latinoamérica, el uso de términos relacionados como Machine Learning, Análisis de datos e inteligencia artificial, es útil a la hora de indagar respecto al mercado laboral debido a la falta de estandarización y sistematización concerniente al campo de acción y áreas laborales del científico de datos. 

A pesar de ésto, de acuerdo a Solís (2021), los perfiles de científico de datos provienen principalmente del mundo de la informática y las telecomunicaciones, con características diversas dependiendo del campo de acción, siendo el dominio de los lenguajes Python y R los más comunmente solicitados, además del manejo de herramientas de visualización. En un estudio realizado por Shaik (2020) en el que analizó más de 1,000 trabajos recientes de científicos de datos haciendo uso de herramientas de Web Scraping para el mercado laboral de la India, encontró que si bien los requerimientos son diversos, el manejo de herramientas de Machine learning, Estadística, análisis y limpieza de datos  además de técnicas de aprendizaje profundo y conocimiento de la nube como AWS, Azure o GCP, y herramientas de visualización, como Tableau y Power Bi, se presentaban como los factores más importantes a la hora de conseguir un empleo como científico de datos. Por otro lado, específicamente respecto al mercado latinoamericano,  Ramírez  (2018) señala que el manejo de lenguajes como Python y R se mantienen como los más populares y los estudios de posgrado son más comunes que en el resto de profesionales relacionados a la computación, siendo el más común el título de maestría. 

Con relación a los factores que contribuyen a la obtención de un mejor salario, Nagpakur (2020)  señala que lo que presenta mayor correlación con un pago más alto son los años de experiencia, más que los grados académicos en sí,(aunque cabe resaltar que éstos se relacionan a los años de experiencia pues implican el tiempo invertido en el desarrollo en el campo de conocimiento específico) lo que se refleja en el aumento salarial conforme aumentan los años de experiencia, por tanto, las personas que no tienen títulos universitarios tienden a tener un mejor salario en términos de pago promedio ya que tienden a acumular más experiencia laboral. Respecto a los puestos con mejor paga, los arquitectos de datos y aquellos que se encuentran en los puestos gerenciales son los que perciben sueldos más altos. 

Por último, cabe resaltar que según éste estudio de 2020,  los países con mejores oportunidades para desarrollarse en éste ámbito son Estados Unidos, Nueva Zelanda, Australia, Alemania  y Canadá; respecto a latinoamérica, si bien México no es el país con más desarrollo (Siendo Brasil quien encabeza la lista) es el país que presenta las mayores ambiciones en temas de IA, siendo el primer país de Latinoamérica que formula una estrategia nacional de IA en el 2018 con objetivos clave para reducir la corrupción y la delincuencia, así como mejorar la salud pública y finalmente impulsar la inclusión financiera (Solís, 2021), sin embargo, de acuerdo a Ramírez (2018), Chile se posiciona como el país más interesante para trabajos en ciencia de datos, presentando el mayor número de vacantes disponibles en relación a su población y uno de los salarios más apropiados en relación al costo de vida del país. 

A pesar de que existen diversas propuestas muy ricas en el anális de datos que se han hecho anteriormente, no se ha encontrado alguna propuesta que se centre en México ni en el análisis de las características específicas de éstos puestos para nuestro mercado laboral, por lo que nuestro objetivo se centrará en explorar las oportunidades de desarrollo para nuestro país. 


**Objetivo**

El objetivo de ésta investigación será el de indagar a través de las herramientas de análisis de datos con Python cuáles son los requerimientos específicos, habilidades, nivel de experiencia, así como el nivel academico necesario para desarrollarse exitosamente en ciencia de datos y comprender cómo ésto se relaciona con las expectativas salariales para ayudar a la toma de decisiones respecto a objetivos profesionales en el camino STEM.

**Preguntas**. 

- ¿Cuál es el salario promedio de un científico de datos?
- ¿Cuántas habilidades en términos de software son requeridas en el ambito laboral?
- ¿Cuál es el área (industria) que paga mejor o que ofrece mejores ofertas?
- ¿Cuáles son los mejores puestos laborales de acuerdo al salario?
- ¿Cuál es el promedio salarial de los distintos puestos laborales?
- ¿Qué sectores tienen mayor interés en un científico de datos?
- ¿Qué país, en promedio, ofrece mejores ofertas salariales?

**Metodología:**

Mediante técnicas de Web Scrapping, se extrajo la información de uno de los portales más populares de ofertas de empleo en tecnología: Glassdoor. Estos datos representan una muestra de las ofertas laborales publicadas en la página Glassdoor entre el 12 y 15 de septiembre del 2022 para los países México, Estados Unidos y Canadá (estos dos últimos por ser los mejores países de nuestro continente para el ejercicio de la ciencia de datos y los cuales ofrecen mejores percepciones salariales). Para esto, se tomó como base el código proporcionado en este [enlace](https://github.com/Atharva-Phatak/Analysing-Glassdoor-Jobs/blob/master/Scrape_data_Final.ipynb) y se realizaron diversas modificaciones para obtener la muestra de la información de las ofertas laborales.


**Operacionalización de variables**: 

+ Científico de datos: Basandonos en la metodología  de Ramírez, (2018), y debido a que no existe un concenso respecto al puesto, usaremos términos relacionados como Machine Learning, análisis de datos e inteligencia artificial, etc., para la obtención en glassdoor. 

+ Nivel de experiencia clasificado a partir de Jr, Semisenior, Senior y Trainee

+ Antecedentes académicos: 
    - Por nivel: Educación básica, licenciatura, posgrado. 

+ Habilidades:
    *   Lenguajes de programación: Python, R, SQL, MongoDB, MariaDB, Azure, Docker, Kubernetes, Java, Julia, Scala, C, C\+\+, JavaScript, Swift, Go, MATLAB, SAS, SPSS, Bash, Visual Basic, PHP, HTML, Ruby, Git, Excel, Powepoint, Word, Tablue, MicroStrategy, PowerBI, Apache Spark y Hadoop


+ Áreas para el ejercicio de ciencia de datos (giro de la empresa)
    -   Industria
    -   Investigación
    -   Educación


## Limpieza de la información

Al realizar el web scrapping, se obtuvieron distintos conjuntos de datos para cada término (palabra clave para el puesto: "Cientifico de datos", etc) y 
estos se pueden consultar en está [página](https://github.com/CarlosFernandoVG/ProyectoPythonBEDU/blob/main/WebScrappingGlassdoor.ipynb) y en la siguiente celda los obtenemos de manera rápida para, posteriormente, unir cada conjunto de datos en uno solo.

In [1]:
import os
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
import numpy as np
import re

# folder path
dir_path_EUA = 'Datasets/EUA'
dir_path_MX = 'Datasets/Mexico'
dir_path_CAN = 'Datasets/Canadá'

# Obtenemos las direcciones de nuestros archivos
dir_path_EUA = os.listdir(dir_path_EUA)
dir_path_MX = os.listdir(dir_path_MX)
dir_path_CAN = os.listdir(dir_path_CAN)

#Colocamos este paso extra, para evitar problemas con las rutas al ejecutar varias veces el notebook
dir_path_EUA = [ruta for ruta in dir_path_EUA if ruta != '.ipynb_checkpoints']
dir_path_MX = [ruta for ruta in dir_path_MX if ruta != '.ipynb_checkpoints']
dir_path_CAN = [ruta for ruta in dir_path_CAN if ruta != '.ipynb_checkpoints']

In [2]:
#Creamos tres data frames con la misma estructura para unir cada conjunto de datos por país

data_EUA = pd.DataFrame(columns = ['Job_title' , 'Company' , 'State' , 'City' , 'Min_Salary' ,'Max_Salary' , 'unitText',
                                      "SalaryCurrency", 'Job_Desc' , 'Industry'  , 'Rating' , 'Date_Posted' , 'Valid_until' , 'Job_Type'])
data_MX = pd.DataFrame(columns = ['Job_title' , 'Company' , 'State' , 'City' , 'Min_Salary' ,'Max_Salary' , 'unitText',
                                      "SalaryCurrency", 'Job_Desc' , 'Industry'  , 'Rating' , 'Date_Posted' , 'Valid_until' , 'Job_Type'])
data_CAN = pd.DataFrame(columns = ['Job_title' , 'Company' , 'State' , 'City' , 'Min_Salary' ,'Max_Salary' , 'unitText',
                                      "SalaryCurrency", 'Job_Desc' , 'Industry'  , 'Rating' , 'Date_Posted' , 'Valid_until' , 'Job_Type'])

In [3]:
#Creamos una lista con los diferntes conjuntos de datos por país
EUA_csvs = []
MX_csvs = []
CAN_csvs = []

for filename in dir_path_EUA:
    df = pd.read_csv("Datasets/EUA/"+filename, index_col=None, header=0)
    EUA_csvs.append(df)
    
for filename in dir_path_MX:
    df = pd.read_csv("Datasets/Mexico/"+filename, index_col=None, header=0)
    MX_csvs.append(df)
    
for filename in dir_path_CAN:
    df = pd.read_csv("Datasets/Canadá/"+filename, index_col=None, header=0)
    CAN_csvs.append(df)

#En este paso, ya tenemos los conjutos de datos unidos por país
EUA_data = pd.concat(EUA_csvs, axis=0, ignore_index=True)
MX_data = pd.concat(MX_csvs, axis=0, ignore_index=True)
CAN_data = pd.concat(CAN_csvs, axis=0, ignore_index=True)

In [4]:
#Quitamos duplicados y reiniciamos índices
EUA_data = EUA_data.drop_duplicates()
EUA_data.reset_index(inplace=True, drop=True)

MX_data = MX_data.drop_duplicates()
MX_data.reset_index(inplace=True, drop=True)
MX_data.drop('Unnamed: 0', inplace=True, axis=1)

CAN_data = CAN_data.drop_duplicates()
CAN_data.reset_index(inplace=True, drop=True)

#Agregamos una columna que indique el país
EUA_data["Country"] = "EUA"
MX_data["Country"] = "MX"
CAN_data["Country"] = "CAN"

Ya que tenemos todos los datos cargados, los compilamos en un solo data frame

In [5]:
jobs_data = pd.concat([EUA_data, MX_data, CAN_data], axis=0, ignore_index=True)
jobs_data.head()

Job_title          Company       State         City  \
0  Data Engineer with Azure Synapse     4urHires INC         NaN          NaN   
1                     Data Engineer       MARS Group         NaN          NaN   
2       Data Engineer - ENTRY LEVEL  PCS Global Tech  California  Los Angeles   
3                 Sr. Data Engineer            Optiv      Kansas      Leawood   
4   Entry-Level Azure Data Engineer       Antra, Inc         NaN          NaN   

  Min_Salary Max_Salary unitText SalaryCurrency  \
0         -1         -1      NaN            USD   
1         -1         -1      NaN            USD   
2         -1         -1      NaN            USD   
3         -1         -1      NaN            USD   
4         -1         -1      NaN            USD   

                                            Job_Desc  \
0  Azure EngineerStrong experience in SynapseJob ...   
1  Description: Data is a critical driver of this...   
2  Responsibilities· Analyze and organize raw dat...   
3  At Optiv, we’re on a mission to help our clien...   
4  About usAntra, Inc is a leading IT consultancy...   

                        Industry Rating          Date_Posted Valid_until  \
0                            NaN    NaN  2022-09-16T00:00:00  2022-10-16   
1                            NaN    NaN  2022-09-14T00:00:00  2022-10-16   
2  Tecnologías de la información    NaN  2022-09-15T00:00:00  2022-10-16   
3  Tecnologías de la información    NaN  2022-09-15T00:00:00  2022-10-16   
4  Tecnologías de la información    NaN  2022-09-13T00:00:00  2022-10-16   

    Job_Type Country  
0      OTHER     EUA  
1  FULL_TIME     EUA  
2  FULL_TIME     EUA  
3  FULL_TIME     EUA  
4  FULL_TIME     EUA

Nuestro conjunto de datos, en este momento, contiene 15 variables y 2576 observaciones:

In [6]:
jobs_data.shape

(2576, 15)

Las columnas tienen el siguiente significado:

- Job_title: Cargo ofrecido por la compañía
- Company: Empresa que ofrece el trabajo
- State / City / Country: Descripción de la ubicación para laborar (estado, ciudad y país)
- Min_Salary / Max_Salary: Estimación oficial del salario base (rango) que pretende pagar el empleador por este trabajo
- SalaryCurrency: Moneda a la cual se pagará el trabajo
- unitTex: Moneda en la cual se encuentra la estimación del rango salarial
- Job_Desc: Descripcción del trabajo
- Industry: Área a la que pertenece la compañia
- Rating: Calificación promedio que los empleados dan a la empresa
- Date_Posted: Fecha en la que se publico la oferta laboral
- Valid_until: Fecha límite para aplicar a la oferta laboral
- Job_Type: Tipo de trabajo (remoto, presencial, etc)

Podemos visualizar una descripción de los tipos de datos que se encuentran en el data frame

In [7]:
jobs_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2576 entries, 0 to 2575
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Job_title       2576 non-null   object
 1   Company         2570 non-null   object
 2   State           2029 non-null   object
 3   City            2029 non-null   object
 4   Min_Salary      2576 non-null   object
 5   Max_Salary      2576 non-null   object
 6   unitText        446 non-null    object
 7   SalaryCurrency  2576 non-null   object
 8   Job_Desc        2569 non-null   object
 9   Industry        1785 non-null   object
 10  Rating          0 non-null      object
 11  Date_Posted     2576 non-null   object
 12  Valid_until     2576 non-null   object
 13  Job_Type        2576 non-null   object
 14  Country         2576 non-null   object
dtypes: object(15)
memory usage: 302.0+ KB


Podemos notar que todos los tipos de datos que contiene nuestro data set es de tipo `object`. Más adelante se modificarán estos tipos de datos por otros más adecuados

Lo anterior nos ayuda a visualizar cuantos valores perdidos se tienen por columna, por lo que procedemos a visualizar la proporción de valores perdidos para determinar si algunas columnas pueden ser eliminadas

In [8]:
jobs_data.isna().sum()/len(jobs_data)

Job_title         0.000000
Company           0.002329
State             0.212345
City              0.212345
Min_Salary        0.000000
Max_Salary        0.000000
unitText          0.826863
SalaryCurrency    0.000000
Job_Desc          0.002717
Industry          0.307065
Rating            1.000000
Date_Posted       0.000000
Valid_until       0.000000
Job_Type          0.000000
Country           0.000000
dtype: float64

`unitTex` contiene un 83% de valores perdidos; eliminaremos esta columna ya que esta indica la moneda de los salarios estimados por glassdoor y en la columna `SalaryCurrency`tenemos la misma información y sin valores perdidos

In [9]:
print(f'Valores perdidos de unitText: {jobs_data["unitText"].isna().sum()}')
print(f'Valores perdidos de SalaryCurrency: {jobs_data["SalaryCurrency"].isna().sum()}')

Valores perdidos de unitText: 2130
Valores perdidos de SalaryCurrency: 0


La columna `Rating` solo contiene información perdida, así que será eliminada

In [10]:
jobs_data.drop(["unitText", "Rating"], inplace=True, axis = 1)
jobs_data.head()

Job_title          Company       State         City  \
0  Data Engineer with Azure Synapse     4urHires INC         NaN          NaN   
1                     Data Engineer       MARS Group         NaN          NaN   
2       Data Engineer - ENTRY LEVEL  PCS Global Tech  California  Los Angeles   
3                 Sr. Data Engineer            Optiv      Kansas      Leawood   
4   Entry-Level Azure Data Engineer       Antra, Inc         NaN          NaN   

  Min_Salary Max_Salary SalaryCurrency  \
0         -1         -1            USD   
1         -1         -1            USD   
2         -1         -1            USD   
3         -1         -1            USD   
4         -1         -1            USD   

                                            Job_Desc  \
0  Azure EngineerStrong experience in SynapseJob ...   
1  Description: Data is a critical driver of this...   
2  Responsibilities· Analyze and organize raw dat...   
3  At Optiv, we’re on a mission to help our clien...   
4  About usAntra, Inc is a leading IT consultancy...   

                        Industry          Date_Posted Valid_until   Job_Type  \
0                            NaN  2022-09-16T00:00:00  2022-10-16      OTHER   
1                            NaN  2022-09-14T00:00:00  2022-10-16  FULL_TIME   
2  Tecnologías de la información  2022-09-15T00:00:00  2022-10-16  FULL_TIME   
3  Tecnologías de la información  2022-09-15T00:00:00  2022-10-16  FULL_TIME   
4  Tecnologías de la información  2022-09-13T00:00:00  2022-10-16  FULL_TIME   

  Country  
0     EUA  
1     EUA  
2     EUA  
3     EUA  
4     EUA

Existen muy pocas observaciones sin descripcción, por lo que procederemos a eliminarlas

In [11]:
jobs_data.dropna(subset = ["Job_Desc"], inplace=True)

In [12]:
jobs_data.reset_index(inplace=True, drop=True)

Podemos imputar de manera sencilla algunas variables:
- Algunas empresas no proporcionan su nombre por motivos de confidencialidad, por lo que imputaremos este valor en la columna `Company`
- Diferentes empresa no especifican su giro, por lo que imputaremos el valor "not specified" en la columna `Industry`

In [13]:
values = {"Company": "confidential", "Industry": "not specified"}
jobs_data = jobs_data.fillna(value=values)

In [14]:
jobs_data.isna().sum()/len(jobs_data)

Job_title         0.000000
Company           0.000000
State             0.212145
City              0.212145
Min_Salary        0.000000
Max_Salary        0.000000
SalaryCurrency    0.000000
Job_Desc          0.000000
Industry          0.000000
Date_Posted       0.000000
Valid_until       0.000000
Job_Type          0.000000
Country           0.000000
dtype: float64

Las columnas `Date_Posted` y `Valid_until` son fechas, por lo que podemos modificar el tipo de dato en dicha variable

In [15]:
jobs_data['Date_Posted'] = pd.to_datetime(jobs_data['Date_Posted'])
jobs_data['Valid_until'] = pd.to_datetime(jobs_data['Valid_until'])

In [16]:
jobs_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2569 entries, 0 to 2568
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Job_title       2569 non-null   object        
 1   Company         2569 non-null   object        
 2   State           2024 non-null   object        
 3   City            2024 non-null   object        
 4   Min_Salary      2569 non-null   object        
 5   Max_Salary      2569 non-null   object        
 6   SalaryCurrency  2569 non-null   object        
 7   Job_Desc        2569 non-null   object        
 8   Industry        2569 non-null   object        
 9   Date_Posted     2569 non-null   datetime64[ns]
 10  Valid_until     2569 non-null   datetime64[ns]
 11  Job_Type        2569 non-null   object        
 12  Country         2569 non-null   object        
dtypes: datetime64[ns](2), object(11)
memory usage: 261.0+ KB


Analizando diversas muestras de aquellas observaciones que contienen valores nulos en las columnas `State` y `City`, podemos notar que muchos trabajos no colocan en donde se trabajaría, ya que pueden ser trabajos por vía remota o prefiere la misma empresa no proporcionar dicha información más que mencionar el país.

Por tal motivo, remplazaremos estos valores por un indicador de omisión de esta información

In [17]:
jobs_data = jobs_data.fillna(value={'State': 'not specified', 'City': 'not specified'})
jobs_data.isna().sum()/len(jobs_data)

Job_title         0.0
Company           0.0
State             0.0
City              0.0
Min_Salary        0.0
Max_Salary        0.0
SalaryCurrency    0.0
Job_Desc          0.0
Industry          0.0
Date_Posted       0.0
Valid_until       0.0
Job_Type          0.0
Country           0.0
dtype: float64

En las columnas `Min_Salary` y `Max_Salary` se tienen valores igual a -1, lo cual representa que la página de glassdoor, para ciertos empleos, no tiene el salario, o el rango salarial, en un apartado especial de la página. Reemplazamos esos valores por `NaN` y determinamos la proporción de valores perdidos

In [18]:
jobs_data['Min_Salary'] = jobs_data['Min_Salary'].replace(-1,np.NaN)
jobs_data['Max_Salary'] = jobs_data['Max_Salary'].replace(-1,np.NaN)

In [19]:
jobs_data.isna().sum()/len(jobs_data)

Job_title         0.00000
Company           0.00000
State             0.00000
City              0.00000
Min_Salary        0.82717
Max_Salary        0.82717
SalaryCurrency    0.00000
Job_Desc          0.00000
Industry          0.00000
Date_Posted       0.00000
Valid_until       0.00000
Job_Type          0.00000
Country           0.00000
dtype: float64

Vamos a analizar la información de la variable `Job_Desc` para obtener más información del salario y poder imputar información en las columnas `Min_Salary`y `Max_Salary`

Como primer intento, notamos que, en caso de estar descrito el salario en la oferta laboral, muchos de estos salarios tienen un formato \\$xxx,xxx.xxx - \\$xxx,xxx.xxx, por lo que los obtenemos con el uso de expresiones regulares

In [20]:
rangos_salariales_1 = jobs_data['Job_Desc'].str.findall(r'(\$\d+[\.\,]\d+[\.\,]*\d*)', flags=re.I)
rangos_salariales_1[0]

['$55.00', '$60.00']

Creamos algunas funciones auxiliares para limpiar y obtener los datos

In [21]:
#Funciones auxiliares:
#Función para eliminar algunos caracteres en las listas que contiene rangos_salariales_1
def fixed_salary(lista):
    if isinstance(lista, str):
        return []
    if(len(lista)> 0):
        for i in range(len(lista)):
            lista[i] = lista[i].replace(",", "")
            lista[i] = lista[i].replace("$", "")
            #Existen algunos casos donde se utilizan los puntos como separadores de miles y decimales a la vez, por ejemplo: $1.425.870
            if lista[i].count(".")>1:
                nums_sep = lista[i].split(".")
                lista[i] = "".join(nums_sep[:len(nums_sep)-1]) + "."+nums_sep[-1]
        return lista
    else:
        return []
#Función para obtener los datos de las listas de rangos_salariales_1 para los salarios mínimos y máximos
def get_salary(lista, index):
    if(len(lista)> 0): 
        #En este caso podemos sacar salario mínimo y máximo
        if len(lista) > 1:
            #Puede suceder que la segunda cifra no sea un salario, por ejemplo: "Salario: $1.425.870 + $117.172 Auxilio de trasporte"
            if float(lista[0])<=float(lista[1]): #Aquí no habria problema
                return float(lista[index])
            else: #Aquí tendríamos que checar
                if index == 0:
                    return float(lista[index])
                else: 
                    return np.NaN
        else: 
            if index > 0:
                return np.NaN
            else:
                return float(lista[0])
    else:
        return np.NaN

Primero limpiamos el texto de el anterior series eliminando algunos caracteres

In [22]:
rangos_salariales_1 = rangos_salariales_1.map(fixed_salary)
rangos_salariales_1.head()

0          [55.00, 60.00]
1          [39.68, 60.00]
2    [60000.00, 70000.00]
3                      []
4    [62000.00, 70000.00]
Name: Job_Desc, dtype: object

Obtenemos el salario mínimo y máximo

In [23]:
aux_Min_Salary = rangos_salariales_1.map(lambda x : get_salary(x, 0))
aux_Max_Salary = rangos_salariales_1.map(lambda x : get_salary(x, 1))
aux_Min_Salary

0          55.00
1          39.68
2       60000.00
3            NaN
4       62000.00
          ...   
2564         NaN
2565         NaN
2566         NaN
2567        5.60
2568         NaN
Name: Job_Desc, Length: 2569, dtype: float64

Ahora fundimos las columnas `Min_Salary` y `Max_Salary` con los series que creamos:

In [24]:
def merge_salary_series(aux_serie, original_serie):
    aux_list = []
    for i in range(len(aux_serie)):
        #Daremos preferencia al rango salarial obtenido por la descripcción del trabajo
        if not np.isnan(aux_serie[i]):
            aux_list.append(aux_serie[i])
        elif not not np.isnan(original_serie[i]):
            aux_list.append(original_serie[i])
        else:
            aux_list.append(np.NaN)
    return(aux_list)

In [25]:
jobs_data['Min_Salary'] = pd.Series(merge_salary_series(aux_Min_Salary, jobs_data['Min_Salary']))
jobs_data['Max_Salary'] = pd.Series(merge_salary_series(aux_Max_Salary, jobs_data['Max_Salary']))

In [26]:
jobs_data['Max_Salary'].head()

0       60.0
1       60.0
2    70000.0
3        NaN
4    70000.0
Name: Max_Salary, dtype: float64

Los salarios máximos y mínimos pueden estar en diferentes formas de pago, por lo que vamos a obtener la temporalidad de pago de los salarios anteriores

In [27]:
#Tenemos diferentes valores que no siguen un patron específico, por lo que optaremos por sustituir todos los casos
different_salaries = {'per hourSchedule': 'hour',
    'per yearSchedule': 'year',
    'not specified': 'not specified',
    'per yearBenefits': 'year',
    'per hourBenefits': 'hour',
    'per hourExperience': 'hour',
    'per monthLanguage': 'month',
    'al mesHorario': 'month',
    'per monthExperience': 'month',
    'per monthBenefits': 'month',
    'al mesPrestaciones': 'month',
    'a la semanaHorario': 'week',
    'al mesEducación': 'month',
    'per monthEducation': 'month',
    'al mesExperiencia': 'month',
    'mensuales brutosModalidad de trabajo de manera HibridaHorarios': 'month',
    'Brutos\nPrestaciones de Ley\nVales de despensa\nEsquema de trabajo hibrido\nZona de trabajo': 'month',
    'de Bono por cumplimiento de objetivosPrestaciones de LeySeguro de vidaCrecimiento y estabilidad laboral\nLugar de trabajo': 'month',
    'al díaHorario': 'day',
    'MXN Mensual\n\nCategoría': 'month',
    'al mes brutosTipo de puesto': 'month',
    'al añoHorario': 'year',
    'mensuales BRUTOS\nPrestaciones de ley\nBonos por asistencia y puntualidad\nUtilidades\nHorario laboral': 'month',
    'al mesIdioma': 'month',
    'netosTipo de puesto': 'month',
    'netos mensualesPrestaciones de LeyPrestaciones superioresLugar de trabajo': 'month',
    'brutos mensualesPrestaciones de ley y superioresOpción a plantaHorario': 'month',
    'mensuales brutosPrestaciones de leyEstabilidad LaboralUniformesTipo de puesto': 'month',
    'mensuales brutoPrestaciones de leyTrabajo de lunes a sábado con descanso en domingoEstabilidad LaboralUniformesTipo de puesto': 'month',
    'al mesEducation': 'month',
    'al mesCrecimiento profesional y económicoCapacitaciones y certificaciones constantesCalificaciones básicas': 'month',
    'Pesos\n\n\nPais': 'month',
    'pesos brutos mensuales que se pagarán de la siguiente forma': 'month',
    'per monthExpected Start Date': 'month',
    'per yearFlexible Language Requirement': 'year',
    'Grade 10 BU\n\nApplication Deadline': 'month',
    'per monthFlexible Language Requirement': 'month',
    'per hour\n\nApplication Close': 'hour',
    'Hourly\nGrade': 'hour',
    'per annum\n\nApplication Close': 'year',
    'per hourFlexible Language Requirement': 'hour',
    'per year plus benefits and paid OT\n\nHow to Apply': 'year',
    'per hourEducation': 'hour'}

def get_salary_scheme(lista):
    if isinstance(lista, str):
        return("not specified")
    if len(lista)>0:
        return(different_salaries[lista[0]])
    else: 
        return("not specified")

Obtenemos y limpiamos las formas de pago

In [28]:
salary_scheme_series = jobs_data['Job_Desc'].str.findall(r'\$\d+[\.\,]\d+[\.\,]*\d* ([\w\s]+):')
salary_scheme_series[salary_scheme_series.isna()] = "not specified"

Finalmente, agregamos el esquema salarial

In [29]:
jobs_data['Salary_Scheme'] = salary_scheme_series.map(get_salary_scheme)

Realizando una busqueda exhaustiva, podemos notar que existen casos que no fueron detectados por el código anterior. Al ser casos puntuales, agregamos estos datos de manera manual

In [30]:
jobs_data['Min_Salary'].loc[47] = 125
jobs_data['Max_Salary'].loc[47] = 145
jobs_data['Salary_Scheme'].loc[47] = "hour"
jobs_data['Min_Salary'].loc[174] = 120000
jobs_data['Max_Salary'].loc[174] = 140000
jobs_data['Salary_Scheme'].loc[174] = "year"
jobs_data['Min_Salary'].loc[197] = 132000
jobs_data['Max_Salary'].loc[197] = 196000
jobs_data['Salary_Scheme'].loc[197] = "year"
jobs_data['Min_Salary'].loc[340] = 120000
jobs_data['Max_Salary'].loc[340] = 180000
jobs_data['Salary_Scheme'].loc[340] = "year"
jobs_data['Min_Salary'].loc[390] = 29000
jobs_data['Max_Salary'].loc[390] = 80000
jobs_data['Salary_Scheme'].loc[390] = "month"
jobs_data['Min_Salary'].loc[442] = 15000
jobs_data['Max_Salary'].loc[442] = 16000
jobs_data['Salary_Scheme'].loc[442] = "month"
jobs_data['Min_Salary'].loc[722] = 65000
jobs_data['Max_Salary'].loc[722] = 75000
jobs_data['Salary_Scheme'].loc[722] = "month"
jobs_data['Min_Salary'].loc[727] = 65000
jobs_data['Max_Salary'].loc[727] = 75000
jobs_data['Salary_Scheme'].loc[727] = "month"
jobs_data['Min_Salary'].loc[734] = 65000
jobs_data['Max_Salary'].loc[734] = 75000
jobs_data['Salary_Scheme'].loc[734] = "month"
jobs_data['Min_Salary'].loc[738] = 65000
jobs_data['Max_Salary'].loc[738] = 75000
jobs_data['Salary_Scheme'].loc[738] = "month"
jobs_data['Min_Salary'].loc[761] = 35000
jobs_data['Max_Salary'].loc[761] = 45000
jobs_data['Salary_Scheme'].loc[761] = "month"
jobs_data['Min_Salary'].loc[767] = 9000
jobs_data['Max_Salary'].loc[767] = 12000
jobs_data['Salary_Scheme'].loc[767] = "month"
jobs_data['Min_Salary'].loc[875] = 57600
jobs_data['Max_Salary'].loc[875] = 128400
jobs_data['Salary_Scheme'].loc[875] = "month"
jobs_data['Max_Salary'].loc[1136] = 18000
jobs_data['Max_Salary'].loc[1136] = 23000
jobs_data['Salary_Scheme'].loc[1136] = "month"
jobs_data['Min_Salary'].loc[712] = 18499.00
jobs_data['Salary_Scheme'].loc[712] = "month"
jobs_data['Min_Salary'].loc[744] = 20000
jobs_data['Salary_Scheme'].loc[744] = "month"
jobs_data['Min_Salary'].loc[759] = 7000
jobs_data['Salary_Scheme'].loc[759] = "month"
jobs_data['Min_Salary'].loc[1186] = 20000
jobs_data['Salary_Scheme'].loc[1186] = "month"
jobs_data['Min_Salary'].loc[1394] = 18499.00
jobs_data['Salary_Scheme'].loc[1394] = "month"

Ahora, podemos ver como disminuyo el porcentaje de valores perdidos del rango salarial

In [31]:
jobs_data.isna().sum()/len(jobs_data)

Job_title         0.000000
Company           0.000000
State             0.000000
City              0.000000
Min_Salary        0.695991
Max_Salary        0.777345
SalaryCurrency    0.000000
Job_Desc          0.000000
Industry          0.000000
Date_Posted       0.000000
Valid_until       0.000000
Job_Type          0.000000
Country           0.000000
Salary_Scheme     0.000000
dtype: float64

Aquellas observaciones con valores nulos nos pueden ayudar a obtener más información acerca de otros temas por lo que eliminar dichas observaciones no sería lo más adecuado. Además, diversas ofertas laborales prefieren omitir el rango salarial por motivos estratégicos, por lo que están justificados estos valores perdidos. En caso de querer aplicar algún modelo u obtener alguna estadística, considerando esta variable,  podemos filtrar los datos por aquellos donde se tenga observaciones completas obteniendo un conjunto de datos con 571 observaciones

In [32]:
jobs_data.dropna().shape

(571, 14)

#  Creación y modificación de variables

De la variable `Job_Desc` podemos obtener mucha información; por ejemplo, podemos obtener cuantos requerimientos, a nivel de software, es solicitado por los empleadores. En este caso buscamos los programas más comúnes en el ámbito de la ciencia de datos

In [33]:
common_software = ["Python", "SQL", "MongoDB", "MariaDB", "Azure", "Docker", "Kubernetes", "Java", "Julia", "Scala", "C", "C\+\+", "JavaScript", "Swift", "Go", "MATLAB", "SAS", "SPSS", "Bash", "Visual Basic", "PHP", "HTML", "Ruby", "Git", "Excel", "Powepoint", "Word", "Tablue", "MicroStrategy", "PowerBI", "Apache Spark", "Hadoop"]

def count_software(text):
    n = 0
    for software in common_software:
        search = re.search(software, text, flags= re.I)
        if not search is None:
            n +=1 
    return n

In [34]:
jobs_data['cantidad_software'] = jobs_data['Job_Desc']. map(count_software)

Para el caso del lenguaje de programación R, se debe dar un tratamiento especial, ya que el anterior código busca los nombres de manera específica y R se puede confundir como parte de una palabra

In [35]:
#Funció para determinar tenemos aislada la letra R en la descripcción y así saber que se trata del lenguaje de programación
def find_R(lista_texto):
    count_R = 0
    for texto in lista_texto:  
        if texto == "R":
            count_R += 1
    if count_R>0 :
        return 1
    else:
        return 0

In [36]:
r_lenguaje_n = jobs_data["Job_Desc"].str.replace(",|\.|\;|\:|\+|·", " ", regex=True).str.split(" ").map(find_R)

Finalmente, añadimos las ocurrencias del lenguaje R a la variable que contaba el software

In [37]:
jobs_data['cantidad_software'] = jobs_data['cantidad_software'] + r_lenguaje_n

Podemos determinar que tipo de experiencia/habilidades (senior, junior, etc.) que están buscando los empleadores

In [38]:
#Obtenemos un Series donde se recolecten los términos junior, senior, trainee, entry level y semi sr
level_data_scientist = jobs_data["Job_title"].str.findall("Junior|Jr|Senior|Sr|Trainee|ENTRY.*LEVEL|Semi.*Sr", flags= re.I)

In [39]:
#Hacemos una función para homogeneizar los anteriores términos
def check_levels(level_list):
    level_list = list(set(level_list))
    if len(level_list) == 0:
        return "Not specified"
    elif len(level_list) > 1:
        return "free level"
    else:
        if level_list[0].lower() in ["junior", "jr", "entry-level", "entry level"]:
            return "Junior"
        elif level_list[0].lower() in ["senior", "sr"]:
            return "Senior"
        else:
            return "SemiSr"

Agregamos una columna con dicha información:

In [40]:
jobs_data['level_required'] = level_data_scientist.map(check_levels)

Hacemos una variable que indique si es requerido algún tipo de grado específico:

In [41]:
jobs_data['Education_required'] = (
    jobs_data["Job_Desc"].str.contains("Degree|Bachelor|licenciatura|Posgrado|Carrera|Grado|Educaci[ó|o]n|Education|Estudios|Studies", flags=re.I).
    apply(lambda x: 1 if x else 0)
)

Tenemos un rango salarial en algunas ofertas, así que obtenemos el promedio del salario para cada oferta laboral donde se tenga la información

In [42]:
jobs_data["Mean_salary"] = (jobs_data['Min_Salary'] + jobs_data['Max_Salary'])/2

Con estas nuevas variables, tenemos la siguiente estructura en nuestro conjunto de datos

In [43]:
jobs_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2569 entries, 0 to 2568
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Job_title           2569 non-null   object        
 1   Company             2569 non-null   object        
 2   State               2569 non-null   object        
 3   City                2569 non-null   object        
 4   Min_Salary          781 non-null    float64       
 5   Max_Salary          572 non-null    float64       
 6   SalaryCurrency      2569 non-null   object        
 7   Job_Desc            2569 non-null   object        
 8   Industry            2569 non-null   object        
 9   Date_Posted         2569 non-null   datetime64[ns]
 10  Valid_until         2569 non-null   datetime64[ns]
 11  Job_Type            2569 non-null   object        
 12  Country             2569 non-null   object        
 13  Salary_Scheme       2569 non-null   object      

Finalmente, modificamos el nombre de nuestras variables

In [44]:
jobs_data.rename(columns={"Job_title": "Puesto laboral", "Company": "Empresa", "State": "Estado", "City" : "Ciudad", "Min_Salary": "Salario mínimo",
                         "Max_Salary": "Salario máximo", "SalaryCurrency": "Tipo de moneda", "Job_Desc": "Descripción", "Industry": "Giro de la empresa",
                         "Date_Posted": "Fecha de publicación", "Valid_until": "Vigencia de la oferta", "Job_Type": "Modalidad", "Country": "País",
                         "Salary_Scheme": "Esquema de pago", "cantidad_software": "Cantidad de lenguajes requeridos", "level_required": "Nivel de experiencia", 
                          "Education_required": "Requerimiento educación", "Mean_salary": "Salario promedio"}, inplace=True)
jobs_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2569 entries, 0 to 2568
Data columns (total 18 columns):
 #   Column                            Non-Null Count  Dtype         
---  ------                            --------------  -----         
 0   Puesto laboral                    2569 non-null   object        
 1   Empresa                           2569 non-null   object        
 2   Estado                            2569 non-null   object        
 3   Ciudad                            2569 non-null   object        
 4   Salario mínimo                    781 non-null    float64       
 5   Salario máximo                    572 non-null    float64       
 6   Tipo de moneda                    2569 non-null   object        
 7   Descripción                       2569 non-null   object        
 8   Giro de la empresa                2569 non-null   object        
 9   Fecha de publicación              2569 non-null   datetime64[ns]
 10  Vigencia de la oferta             2569 non-null 

Finalmente, guardamos nuestro conjunto de datos

In [ ]:
jobs_data.to_csv("glassdoor_jobs.csv")

## Análisis exploratorio

In [45]:
jobs_data_clean = jobs_data.dropna()
jobs_data_clean.reset_index(inplace=True, drop=True)

¿Cuál es el salario promedio de un científico de datos? 
Para determinarlo creamos una función para hacer la conversión a las diferentes denominaciones y obtener el salario promedio de toda la base de datos en dolares para simplificar el calculo del mismo

In [46]:
def standar_salary(mean_salary, currency, scheme):
    standar_money = []
    for i in range(len(mean_salary)): 
        money = mean_salary[i]
        #Convertimos a dolares (USD) para estandarizar nuestros valores
        if currency[i] == "CAD":
            money = money*0.75
        if currency[i] == "MXN":
            money = money*0.050
        #Pasamos todos los salarios a pagos mensuales (vamos a suponer una jornada legal de 8 horas)
        if scheme[i] == "hour":
            money = money*8*5*4 #8 horas por 5 días por 4 semanas
        if scheme[i] == "year":
            money = money/12
        if scheme[i] == "week":
            money = money*4
        standar_money.append(money)
    return(standar_money)

In [47]:
jobs_data_clean["Salario promedio"] = pd.Series(standar_salary(list(jobs_data_clean["Salario promedio"]), list(jobs_data_clean["Tipo de moneda"]), list(jobs_data_clean["Esquema de pago"])))

In [48]:
jobs_data_clean["Salario promedio"].mean()

13557.049052940745

Buscamos identificar la expectativa salarial segun la experiencia y habilidades del analista, obtenemos el salario promedio con base al nivel de experiencia requerido por los empleadores para poder identificar la relación que hay entre las variables.

In [49]:
jobs_data_clean.groupby('Nivel de experiencia').agg({'Salario promedio' :"mean"})

Salario promedio
Nivel de experiencia                  
Junior                    13504.643530
Not specified             11633.753561
SemiSr                      400.000000
Senior                    21931.160569
free level                 3285.416667

Podemos notar que un nivel Senior gana, en promedio, más que cualquier otro nivel

¿Cuántas habilidades en términos de software son requeridas en el ambito laboral?

Mediante el campo de `Cantidad de lenguajes requeridos`, se obtienen datos estadísticos del total de los lenguajes  como lo es el promedio 

In [50]:
jobs_data['Cantidad de lenguajes requeridos'].describe()

count    2569.000000
mean        4.792916
std         2.312252
min         1.000000
25%         3.000000
50%         4.000000
75%         6.000000
max        14.000000
Name: Cantidad de lenguajes requeridos, dtype: float64

Es decir, en promedio se solicitan 4 lenguajes de programación para un empleo de ciencia de datos

Para identificar la cantidad de herramientas (lenguajes requeridos) que se requieren según la experiencia del analista, obtenemos la relación entre `Nivel de experiencia` y `Cantidad de Lenguajes requerido` lo que nos indica que un Científico de Datos Senior siempre será mejor pagado que los demás, en free level aparece de formaindistinta y sin etiquetar el nivel de experiencia del programador debido a que el empleador no lo menciona en la descripción.

In [51]:
jobs_data.groupby('Nivel de experiencia').agg({'Cantidad de lenguajes requeridos' :"mean"})

Cantidad de lenguajes requeridos
Nivel de experiencia                                  
Junior                                        4.440367
Not specified                                 4.577177
SemiSr                                        2.666667
Senior                                        5.433616
free level                                    4.916667

A continuación determinamos cual es la cantidad de salario con base en las habilidades de software que domina el analista con lo que se identifica que entre más lenguajes se dominen mayor es el promedio de ingreso

In [52]:
jobs_data.groupby('Cantidad de lenguajes requeridos').agg({'Salario promedio' :"mean"})

Salario promedio
Cantidad de lenguajes requeridos                  
1                                     12717.046154
2                                     26659.956184
3                                     34684.554295
4                                     37650.390941
5                                     57050.587887
6                                     51694.915286
7                                     58792.101786
8                                     66060.757414
9                                     77474.909091
10                                    72940.624444
11                                    34080.000000
12                                    56018.000000
13                                    54500.000000
14                                             NaN

¿Cuál es el área (industria) que paga mejor o que ofrece mejores ofertas?

Lo estimamos obteniendo la relación entre las variables `Giro de la empres` y `Salario promedi` con lo que se observa que en los medios de comunicación, el sector turístico y de construcción y mantenimiento son los que mejor recompensa brindan a sus empleados

In [53]:
jobs_data_clean.groupby('Giro de la empresa').agg({'Salario promedio' :"mean"}).sort_values("Salario promedio", ascending = False)

Salario promedio
Giro de la empresa                                                  
Audiovisual y medios de comunicación                    86250.000000
Hoteles y complejos turísticos                          81088.875000
Servicios de construcción, reparación y manteni...      75713.560000
Aeroespacial y defensa                                  70537.500000
ONG y Organizaciones sin fines de lucro                 60157.500000
Ventas al mayoreo y al menudeo                          41234.375000
Finanzas                                                39054.891204
Gobierno y administración pública                       36383.334670
Tecnologías de la información                           33115.148130
Seguros                                                 29159.249250
Educación                                               27337.987847
Transporte y logística                                  23299.287500
Manufactura                                             18169.052083
Salud                                                    9967.041346
Administración y consultoría                             6506.592593
not specified                                            5038.061877
Recursos humanos                                         4854.216094
Farmacéutica y biotecnología                              868.138889
Telecomunicaciones                                        780.000000
Restaurantes y servicios de comidas                       550.000000
Energía, minería e infraestructura pública                456.094687

¿Cuáles son los mejores puestos laborales de acuerdo al salario?

Lo determinamos obteniendo la relación entre El tipo de puesto a ocupar y el salario promedio, podemos observar que los puestos mejor remunerados son los de Ingeniero de Software y de Machine Learning.

In [54]:
jobs_data_clean.groupby('Puesto laboral').agg({'Salario promedio' :"mean"}).sort_values(by = "Salario promedio", ascending = False).head(10)

Salario promedio
Puesto laboral                                                      
Software Engineer, Machine Learning - Personali...         212000.00
Software Engineer, .NET, Intelligence Features             175000.00
University Grad Machine Learning Engineer 2023 ...         154500.00
Financial Planning & Analysis Manager                      150904.62
Senior Software Development Engineer                       148312.50
Software Engineer III, Machine Learning, Fitbit            147500.00
Junior/Mid-level Software Engineer - Data Engin...         143700.00
Analytics Engineer, DataSF (1042)                          142155.00
Data Analyst II                                            124000.00
Senior Full Stack Engineer (Toronto or Vancouve...         114375.00

¿Qué sectores tienen mayor interés en un científico de datos?

Al agrupar los empleos según el giro de la empresa logramos estimar que en las empresas que tienen un giro relaciona a las tecnologías de la información y finanzas tienen mas empleabilidad para un científico de datos

In [55]:
jobs_data.groupby('Giro de la empresa').size().reset_index(name='Conteo de trabajos').sort_values("Conteo de trabajos", ascending = False)

Giro de la empresa  Conteo de trabajos
25                                      not specified                 786
21                      Tecnologías de la información                 632
9                                            Finanzas                 216
13                                        Manufactura                 187
0                        Administración y consultoría                 111
7          Energía, minería e infraestructura pública                  75
24                     Ventas al mayoreo y al menudeo                  68
4                Audiovisual y medios de comunicación                  67
8                        Farmacéutica y biotecnología                  60
18                                            Seguros                  55
17                                              Salud                  49
10                  Gobierno y administración pública                  48
22                                 Telecomunicaciones                  44
6                                           Educación                  28
15                                   Recursos humanos                  27
19  Servicios de construcción, reparación y manten...                  26
23                             Transporte y logística                  18
1                              Aeroespacial y defensa                  17
14            ONG y Organizaciones sin fines de lucro                  15
11                     Hoteles y complejos turísticos                  13
5                                       Bienes raíces                   9
3                             Artes y entretenimiento                   5
2                                             Agrario                   5
12                                              Legal                   3
16                Restaurantes y servicios de comidas                   3
20                 Servicios personales al consumidor                   2

¿Qué país, en promedio, ofrece mejores ofertas salariales?

Obtenemos la relación entre los paises y su promedio salarial con lo que identificamos que Estados Unidos  tiene la mejor percepcion salarial para un Científico de Datos seguido de Canada y al final se encuentra México.

In [56]:
jobs_data_clean.groupby('País').agg({'Salario promedio' :"mean"})

Salario promedio
País                  
CAN       26550.433724
EUA       34566.335768
MX         1325.997183

Conclusiones

+ El salario que puede llegar a percibir un científico de datos depende en gran medida del rol que ocupen en la empresa seguido del país de la oferta, de las herramientas que domine el analista como son los lenguajes de programación también influyen en menor medida otros factores como los estudios con los que se cuente y giro de la empresa donde se ingrese a laborar.
    
+ Los niveles de habilidades y experiencia lo estandarizan los empleadores bajo el nombre de Senior, Junior, semiSenior, Trainee para poder estimar los  perfiles de los analistas y con base en esto determinar salarios y también si pueden ser contratados.Los salarios promedio obtenidos son para Junior de $13,504.6 USD mensuales, para Senior de $21,931 USD mensuales y para puestos sin categorizar de $3,285.4 USD y $11,633.75. 
    
+ Las áreas donde hay mejor percepción salarial para los científicos de datos son la industria de los medios de comunicación y el turismo, aunque en cuanto a la cantidad de ofertas laborales se refiere, son los Ingenieros de Software los que tienen más probabilidades de contratarse.
        
+ La ciencia de datos debido a la cantidad de conocimientos en programación, matemáticas y estadistica que maneja tambien se ha agrupado en otros nombres o terminos por los distintos sectores que la requieren para su empleabilidad y ejercicio como Ingeniero de Inteligencia artificial, Ingeniero de Machine Learning, Analista de Datos, etc pero todo esto se resume a un rol de analisis de datos implementando algoritmos de estadística, minería de datos, Machine Learning entre otros. 
    
+ Se pudo determinar que el país con mejor expectativa salarial tiene para un Científico de Datos es Estados Unidos, por lo que es la mejor opción para buscar desarrollarse en esta área
    
    
+ Otra herramienta importante a mencionar es la de conocimiento o experiencia en algoritmos de Machine Learning o de Inteligencia Artificial para determinar el rango salarial que puede aspirar un Científico de Datos la cual es esencial en este campo de conocimiento.
    
+ Se eligió una serie de pregunatas de Investigación a futuro las cuales con los datos obtenidos en el presente analisis no nos brindan suficientes elementos para justificar una respuesta puesto que solo se hizo limpieza y depuración de los datos, sin embargo con la aplicacion de un algoritmo para un analisis posterior es posible obtner mayores resultados y conclusiones por lo que agregan al final del presente  documento.

**Investigación a futuro**:

- ¿Cuáles son las expectativas salariales por nivel de experiencia?
- ¿Cuáles son las expectativas salariales dados los diferentes antecedentes académicos?
- ¿Qué habilidades blandas son más valoradas para la obtención de mejores puestos laborales y un perfil más atractivo?
- ¿Cuáles son las prestaciones con las que cuenta?
- ¿Cuáles son los mejores puestos laborales considerando más factores que solo el salario?
- ¿Qué tipo de trabajo podemos esperar (Home office, medio tiempo, total)?
- ¿Existe alguna edad límite para iniciar en el campo de ciencia de datos?
- ¿Cuál es la probabilidad de que alguien con una carrera distinta pueda migrar a científico de datos?

**Referencias**: 

10 Awesome Jobs in Data Science Right Now (2020) Recuperado de: 
https://www.businessstudent.com/careers/hot-career-opportunities-in-data-science-right-now/

Gouviera, R. (2021). ¿Cuánto gana un Data Scientist en paíse de habla hispana? Recuperado de :  https://gonzalezgouveia.com/cuanto-gana-un-data-scientist-en-paises-de-habla-hispana/?utm_source=rss&utm_medium=rss&utm_campaign=cuanto-gana-un-data-scientist-en-paises-de-habla-hispana

Ramírez, L (2018).  ¿Qué necesitas Saber ANTES de buscar trabajo como Data Scientis en LATAM?. Recuperado de https://medium.com/datos-y-ciencia/qu%C3%A9-necesitas-saber-antes-de-buscar-trabajo-como-data-scientist-en-latam-8d3321196069

Ramírez, P. (2021). Analista de datos: el empleo del futuro por el que las empresas pagan 77.000 €, Recuperado de 
https://www.lainformacion.com/management/empleo-analista-datos-sueldo-alto-trabajo-demandado/2835431/


Shaik, S. (2020). Know what employers are expecting for a data scientist role in 2020. Recuperado de  https://towardsdatascience.com/know-what-employers-are-expecting-for-a-data-scientist-role-in-2020-65ad68553cc4


Solís, L. (2021) Los países más avanzados en ciencia de datos en latinoamérica. Recuperado de:  https://cursos.blen.io/los-paises-mas-avanzados-en-ciencia-de-datos-de-latinoamerica/

Nagpurkar, M (2020). Most lucrative Country to work as a data professional. Recuperado de: https://towardsdatascience.com/most-lucrative-country-to-work-in-as-a-data-professional-b2077f696ccb7
Shaik,  


Tokio School 
https://www.tokioschool.com/noticias/analista-datos-sueldo/